capsules/src/ieee802154/framer.rs

```diff
@@ -843,6 +843,7 @@ impl<'a, M: Mac, A: AES128CCM<'a>> CCMClient for Framer<'a, M, A> {
     fn crypt_done(&self, buf: &'static mut [u8], res: ReturnCode, tag_is_valid: bool) {
         let mut tx_waiting = false;
         let mut rx_waiting = false;
+        let rx_waiting_ref = &mut rx_waiting; //needed to mutate rx_waiting in move closure
```

Comment:
I believe the other way we can do this is changing this closure to no longer be `move` and convincing the closure to capture `buf` explicitly:

```diff
@@ -843,7 +843,6 @@ impl<'a, M: Mac, A: AES128CCM<'a>> CCMClient for Framer<'a, M, A> {
     fn crypt_done(&self, buf: &'static mut [u8], res: ReturnCode, tag_is_valid: bool) {
         let mut tx_waiting = false;
         let mut rx_waiting = false;
-        let rx_waiting_ref = &mut rx_waiting; //needed to mutate rx_waiting in move closure

         // The crypto operation was from the transmission pipeline.
         let opt_buf = if let Some(state) = self.tx_state.take() {
@@ -880,7 +879,8 @@ impl<'a, M: Mac, A: AES128CCM<'a>> CCMClient for Framer<'a, M, A> {

         // The crypto operation was from the reception pipeline.
         if let Some(buf) = opt_buf {
-            self.rx_state.take().map(move |state| {
+            self.rx_state.take().map(|state| {
+                let buf = buf;
                 match state {
                     RxState::Decrypting(info) => {
                         let next_state = if tag_is_valid {
@@ -892,7 +892,7 @@ impl<'a, M: Mac, A: AES128CCM<'a>> CCMClient for Framer<'a, M, A> {
                         self.step_receive_state();
                     }
                     other_state => {
-                        *rx_waiting_ref = match other_state {
+                        rx_waiting = match other_state {
                             RxState::ReadyToDecrypt(_, _) => true,
                             _ => false,
                         };
```

IIUC correctly, the fundamental problem is that `||` and `|move |` choose between referencing captured variables and moving them into the closure respectively, but we want to do one of each. In this case, because `buf` has a `'static` lifetime, which means that `||` can't automatically create a reference to it, which is why this closure was using `move` in the first place. For variables that can't be moved, `|move |` will implicitly `Copy` them if possible. Unfortunately, because `rx_waiting` is used after this closure, it can't safely be moved into the closure, so `move` helpfully implicitly copied it too.

The proposed patch in the PR resolved this by switching `rx_waiting` to a type where the implicit copy is to a reference. This alternative solution eliminates implicit copies (so `rx_waiting` is passed by reference) and adds an explicit copy of `buf`.

I _believe_ these would actually both compile down to the same code in practice. However, I think the implicit copy of `|move |` is more error-prone, and switching to the explicit-copy model here might be safer / clearer in the long run.

[I found this blog post helpful in understanding closure nuances.](https://stevedonovan.github.io/rustifications/2018/08/18/rust-closures-are-hard.html)

Reply:
I can definitely say that the sdcard capsule did not take a nuanced approach. `move` closures are just something I apply when the compiler tells me to in an error message.